### DISCUSSION TITLE
Is is possible to evalute lazy data wrappers on serialization?

### DISCUSSION THREAD

**Discussion by RoiEXLab** [nlohmann/json#3144] [Q&A] [answered]
Hi,
I'm wondering if it would be possible to use this json library to create JSON dumps based on database data, without having to read the whole database into memory first.
I'm thinking about code along those lines:
```c++
#include "json.hpp"
#include <fstream>

using jsonf = nlohmann::json;

int main() {
    jsonf jsonfile;

    jsonfile["foo"] = dbquery(); // lazy operation, database connection is not active

    std::ofstream file("key.json");
    file << jsonfile; // here dbquery() is evaluated for all of its entries, writing each entry individually
}
```

I thought about implementing `to_json` and `from_json` for a custom wrapper object, but from looking at the implementation I don't think this will solve the problem.
Is there anything that can be done to solve this problem or would it be better to implement this kind of serialization manually?

**Comment by nlohmann** [comment, marked as answer]
I'm afraid this is currently not possible as the library cannot store lambdas or function pointers. You may be able to pull this off in a `to_json` method though - but then note function `dbquery()` will then be called during conversion. There is no extension point to execute custom code while serializing.

**Reply by RoiEXLab** [reply]
Already assumed this, but okay I will have to write my own serializer then.
I was secretly hoping there might be a hidden way to do it, but already assumed that there wasn't.
Might be something worth considering for a future release of this library though. I don't think I'm the only one crazy enough to have such a use-case: Assigning an iterable object as a json value without having to ever convert it to access the data it contains.
Only when it is accessed the iterator is actually being used.

**Reply by nlohmann** [reply]
I don't say it wouldn't be useful - but this is a bigger change and needs a good concept.

**Reply by RoiEXLab** [reply]
Agreed, do you think it could be added without introducing a breaking change, or is the internal logic too dependent on existing behaviour? In the former case I would consider digging through the code and try to create a pull request if it seems reasonable

**Reply by nlohmann** [reply]
It is actually a concept issue. Right now, serialization assumes to have the JSON value in memory.
